src/dos/cdrom_image.cpp

```diff
@@ -397,134 +416,201 @@ bool CDROM_Interface_Image::GetMediaTrayStatus(bool& mediaPresent, bool& mediaCh
+	else if (track->attr == 0x40)
+		LOG_MSG("CDROM: Game is trying to play audio from a data track (ignoring)");
 
-	// Checks passed, setup the audio stream
-	else {
-		TrackFile* trackFile = tracks[track].file;
-
-		// Convert the playback start sector to a time offset (milliseconds) relative to the track
-		const Bit32u offset = tracks[track].skip + (start - tracks[track].start) * tracks[track].sectorSize;
-		is_playable = trackFile->seek(offset);
-
-		// only initialize the player elements if our track is playable
-		if (is_playable) {
-			const Bit8u track_channels = trackFile->getChannels();
-			const Bit32u track_rate = trackFile->getRate();
-
-			player.cd = this;
-			player.trackFile = trackFile;
-			player.startRedbookFrame = start;
-			player.totalRedbookFrames = len;
-			player.isPlaying = true;
-			player.isPaused = false;
-
-			if (trackFile->getEndian() == AUDIO_S16SYS) {
-				player.addFrames = track_channels ==  2  ? &MixerChannel::AddSamples_s16 \
-				                                         : &MixerChannel::AddSamples_m16;
-			} else {
-				player.addFrames = track_channels ==  2  ? &MixerChannel::AddSamples_s16_nonnative \
-				                                         : &MixerChannel::AddSamples_m16_nonnative;
-			}
-			// Convert Redbook frames to Track frames, rounding up to whole integer frames.
-			// Round up to whole track frames because the content originated from whole redbook frames,
-			// which will require the last fractional frames to be represented by a whole PCM frame.
-			player.totalTrackFrames = static_cast<Bit32u>(ceil(
-			    static_cast<float>(track_rate * player.totalRedbookFrames) / REDBOOK_FRAMES_PER_SECOND));
-			player.playedTrackFrames = 0;
+	// Guard 5: Bail if our track's file-pointer is empty
+	else if (track->file == nullptr)
+		LOG_MSG("CDROM: The file-pointer associated with track %2d is empty",
+		        track->number);
+
+	// Guard 6: Bail if our player's audio channel is empty
+	else if (player.channel == nullptr)
+		LOG_MSG("CDROM: The CD player's mixer channel hasn't been allocated yet");
+
+	// Finished without hitting any of the above issues? Great!
+	else rcode = true;
+
+	// Guard 7: Check if we failed any of the above..
+	if (rcode == false) {
+		StopAudio();
+		// SDL_UnlockMutex(player.mutex);
+		return rcode;
+	}
+
+	// Convert the requested absolute start sector to a byte offset relative to the track's start
+	// Note: even though 'GetTrack() has determined the requested sector falls within our given
+	//       track, it's still possible that the sector is outside of the "physical" bounds of the
+	//       file itself - such as in the pre-gap region. Therefore, we clamp the offset within the
+	//       bounds of the actual track.
+
+	const int relative_start = start - track->start;
+	// If the request falls in the pregap we deduct the difference from the playback duration.
+	if (relative_start < 0) {
+		len -= relative_start;
```

Comment:
The pregap on physical CD's will count toward the number of played frames, however file-based tracks only have a logical pregap fabricated by dosbox, so our start position is actually "moved ahead" to the start of  the audio-file; so in these cases we deduct any pregap frames from the requested duration to avoid playing further into the file or into the next file. 
(poorly worded.. need a little picture to do it justice :-)